### PR TITLE
[mds-logger] [PLAT-194] Fix & enhance debug logging

### PR DESCRIPTION
**Pull request by avatarneil**
## 📚 Purpose
This PR started out as an attempt to fix our Pino debug logging, and evolved into a total overhaul of mds-logger's debug functionality. mds-logger now supports debug logs using the `debug` module, which allows for fancy tracing, namespacing, etc... See the README for more info. 

This also does a lot of under-the-hood refactoring to lean more heavily on pino's additional key functionality versus our own custom log structuring.

## 👌 Resolves:
`logger.debug()` not working at all

## 📦 Impacts:
basically everything

## ✅ PR Checklist
- [x] Test locally (there's no good way to test in the test suite)

### CODE REVIEW

**Comment by avatarneil** [comment]
Converting to draft because pino-debug is being a PITA, and requires invocation in the node process call itself (e.g. `node -r pino-debug foo.js`)

**Review by nessur — APPROVED** [review]

File: packages/mds-logger/index.ts [hunk 1]
```diff
@@ -47,26 +69,32 @@ const log =
...
+ */
+const debugLog = (namespace?: string) => (message: string, data?: Record<string, unknown> | Error) => {
+  const formattedLogMessage = formatLogMessage('debug', message, data)
+  const { log_message, log_data } = formattedLogMessage
+
+  // Invoke the debug module with custom namespace & formatted message
+  debug(namespace ? `mds:${namespace}` : 'mds')(formattedLogMessage)
+
+  return { log_message, log_data }
+}
```
Comment (drtyh2o) [inline packages/mds-logger/index.ts:93]: I think `debug()` only needs to be invoked one time.

```suggestion
const debugLog = (namespace?: string) => {
  const debugLogger = debug(namespace ? `mds:${namespace}` : 'mds')
  return (message: string, data?: Record<string, unknown> | Error) => {
    const formattedLogMessage = formatLogMessage('debug', message, data)
    const { log_message, log_data } = formattedLogMessage
    // Invoke the debug module with custom namespace & formatted message
    debugLogger(formattedLogMessage)
    return { log_message, log_data }
  }
}
```

**Review by drtyh2o — COMMENTED** [review]
Good start. This will be super helpful.

File: packages/mds-logger/index.ts (hunk 1 above)
Comment (avatarneil) [reply packages/mds-logger/index.ts]: Fixed in a581f89!

File: packages/mds-logger/index.ts [hunk 2]
```diff
@@ -14,59 +14,30 @@
...
   (level: LogLevel) =>
-  (
-    message: string,
-    data?: Record<string, unknown> | Error
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  ): { log_message?: string; log_data?: any } => {
+  (message: string, data?: Record<string, unknown> | Error): void => {
     if (process.env.QUIET === 'true') {
-      return {}
+      return
```
Comment (drtyh2o) [inline packages/mds-logger/index.ts:31]: Now that this is a `void` function, you could just wrap the entire body in:
```ts
if (process.env.QUIET !== 'true') {
  ...
}
```

File: packages/mds-logger/index.ts [hunk 3]
```diff
@@ -14,59 +14,30 @@
...
-      log_ISO_timestamp,
-      log_timestamp,
-      ...(log_requestId ? { log_requestId } : {}),
-      log_message,
-      log_data
-    })
-
-    return { log_message, log_data }
+    if (data) return logger[level](redact(data), message)
+    return logger[level](message)
```
Comment (drtyh2o) [inline packages/mds-logger/index.ts:35]: I don't like `if` statements without `{}`:
```ts
return data ? logger[level](redact(data), message) : logger[level](message)
```

File: packages/mds-logger/loggers.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+export const debugLog = (namespace?: string) => {
+  const debugLogger = debug(namespace ? `mds:${namespace}` : 'mds')
+  return (message: string, data?: Record<string, unknown> | Error) => {
+    const formattedLogMessage = { log_message: message, log_data: redact(data), ...deriveRequestId() }
+    const { log_message, log_data } = formattedLogMessage
+
+    // Invoke the debug module with custom namespace & formatted message
+    debugLogger(formattedLogMessage)
+
+    return { log_message, log_data }
```
Comment (drtyh2o) [inline packages/mds-logger/loggers.ts:47]: I thought you were going to remove this?

**Review by drtyh2o — COMMENTED** [review]
Looking better!

File: packages/mds-logger/loggers.ts (hunk 4 above)
Comment (avatarneil) [reply packages/mds-logger/loggers.ts:47]: I should rename the keys to match what we do when pino logging 👍🏼

Comment (avatarneil) [reply packages/mds-logger/loggers.ts:47]: Jk, ought to remove it entirely :)

Comment (avatarneil) [reply packages/mds-logger/loggers.ts]: Resolved in b199007

File: packages/mds-logger/index.ts (hunk 3 above)
Comment (avatarneil) [reply packages/mds-logger/index.ts]: Resolved in b199007

File: packages/mds-logger/index.ts (hunk 2 above)
Comment (avatarneil) [reply packages/mds-logger/index.ts]: Resolved in b199007

**Review by nessur — APPROVED** [review]

File: packages/mds-logger/jest.config.js [hunk 5]
```diff
@@ -0,0 +1 @@
+module.exports = { ...require('../../.jest/jest.config.js'), coverageThreshold: { global: { lines: 80 } } }
```
Comment (drtyh2o) [inline packages/mds-logger/jest.config.js:1]: This file seems oddly un-formatted.

Comment (avatarneil) [reply packages/mds-logger/jest.config.js:1]: ¯\_(ツ)_/¯ when I run prettier it doesn't do anything

File: packages/mds-logger/README.md [hunk 6]
```diff
@@ -1,3 +1,28 @@
...
+import logger from "@mds-core/mds-logger";
+
+// writing with the default namespace
+logger.debug()("Some super debug message");
+
+// writing with a custom namespace
+logger.debug("agency")("Some super duper debug message from agency!");
+
+// writing out with log data
+logger.debug("agency")("some log message with data", { foo: "bar" });
```
Comment (drtyh2o) [inline packages/mds-logger/README.md:23]: I just noticed this, but typically the way node's `debug` is used is that it's called once to get a logger instance and then used. So, it would typically be used within a module like:
```ts
const agencyDebug = logger.debug('agency')
agencyDebug("Some super duper debug message from agency!")
agencyDebug("some log message with data", { foo: "bar" })
```

I wonder if it might make some sense to create separate loggers. Something like:
`agency-logger.ts`
```ts
import logger from '@mds-core/mds-logger'
export default {
  ...logger,
  debug: logger.debug('agency')
}
```

Then mds-agency modules could:
```ts
import logger from './agency-logger'
```

Something like that anyway. You get the idea.

Comment (avatarneil) [reply packages/mds-logger/README.md:23]: I believe that the currently implemented API for `mds-logger` should support this use-case, yup ✅

Comment (drtyh2o) [reply packages/mds-logger/README.md:23]: I agree. I just think that that should be documented as the preferred pattern. In fact, `mds-logger` could export a factory function that accepts a namespace and returns a logger object, such that an `agency-logger.ts` could simply:
```ts
import { CreateLogger } from '@mds-core/mds-logger'
export default CreateLogger('agency')
```
Good work.

**Review by drtyh2o — APPROVED** [review]
Good stuff here!

File: packages/mds-logger/README.md (hunk 6 above)
Comment (avatarneil) [reply packages/mds-logger/README.md:23]: Definitely. I'll probably add a factory & update the examples when I cut a PR to switch everything over to the namespaced logging methods ✅
